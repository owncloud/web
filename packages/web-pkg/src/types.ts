export interface AncestorMetaDataValue {
  id: string
  shareTypes: number[]
  parentFolderId: string
  spaceId: string
  path: string
}

export type AncestorMetaData = Record<string, AncestorMetaDataValue>

export type FederatedUser = {
  display_name: string
  idp: string
  mail: string
  user_id: string
}

export type FederatedConnection = FederatedUser & {
  id: string
}

export enum MouseEventButton {
  LEFT = 0,
  MIDDLE = 1,
  RIGHT = 2
}
