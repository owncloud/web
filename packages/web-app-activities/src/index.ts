import translations from '../l10n/translations.json'
import { useGettext } from 'vue3-gettext'
import { computed } from 'vue'
import { AppMenuItemExtension, defineWebApplication, Extension } from '@ownclouders/web-pkg'
import { urlJoin } from '@ownclouders/web-client'
import { RouteRecordRaw } from 'vue-router'
import { APPID } from './appid'

export default defineWebApplication({
  setup() {
    const { $gettext } = useGettext()

    const appInfo = {
      name: $gettext('Activities'),
      id: APPID,
      icon: 'pulse',
      color: '#887ef1'
    }

    const routes: RouteRecordRaw[] = [
      {
        path: '/',
        name: 'root',
        component: () => import('./LayoutContainer.vue'),
        redirect: urlJoin(appInfo.id, 'list'),
        meta: {
          authContext: 'user'
        },
        children: [
          {
            path: 'list',
            name: 'list',
            component: () => import('./views/App.vue'),
            meta: {
              authContext: 'user',
              title: $gettext('Activities')
            }
          }
        ]
      }
    ]

    const menuItemExtension: AppMenuItemExtension = {
      id: `app.${appInfo.id}.menuItem`,
      type: 'appMenuItem',
      label: () => appInfo.name,
      color: appInfo.color,
      icon: appInfo.icon,
      priority: 30,
      path: urlJoin(appInfo.id)
    }
    const extensions = computed(() => {
      const result: Extension[] = []

      result.push(menuItemExtension)

      return result
    })

    return {
      appInfo,
      routes,
      translations,
      extensions
    }
  }
})
