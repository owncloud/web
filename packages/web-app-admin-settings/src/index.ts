import translations from '../l10n/translations.json'
import General from './views/General.vue'
import Users from './views/Users.vue'
import Groups from './views/Groups.vue'
import Spaces from './views/Spaces.vue'
import { Ability, urlJoin } from '@ownclouders/web-client'
import {
  ApplicationInformation,
  AppMenuItemExtension,
  AppNavigationItem,
  defineWebApplication,
  useAbility,
  useUserStore
} from '@ownclouders/web-pkg'
import { RouteRecordRaw } from 'vue-router'
import { computed } from 'vue'

// just a dummy function to trick gettext tools
function $gettext(msg: string) {
  return msg
}

const appId = 'admin-settings'

function adminAcrGuard(to, from, next) {
  if (typeof window === 'undefined') {
    next()
    return
  }

  // Avoid re-triggering when returning from step-up or when a request is already in-flight for this route
  const hasStepupParam = !!to.query.stepup
  // const hasGlobalStepupFlag = new URLSearchParams(window.location.search).has('stepupAcr')
  const stepupKey = 'oc.stepup:admin-settings'
  const alreadyRequested = sessionStorage.getItem(stepupKey) === '1'

  // diagnostics
  try {
    console.debug('[adminAcrGuard] enter', {
      to: to.fullPath,
      hasStepupParam,
      // hasGlobalStepupFlag,
      alreadyRequested,
      locationSearch: window.location.search
    })
  } catch (_) {}

  // if (!hasStepupParam && !hasGlobalStepupFlag && !alreadyRequested) {
  if (!hasStepupParam && !alreadyRequested) {
    const redirectUrl = to.fullPath + (to.fullPath.includes('?') ? '&' : '?') + 'stepup=1'
    try {
      sessionStorage.setItem(stepupKey, '1')
    } catch (_) {
      // ignore storage errors
    }
    try {
      console.info('[adminAcrGuard] initiating step-up', { redirectUrl })
    } catch (_) {}
    window.location.assign(`/?stepupAcr=advanced&redirectUrl=${encodeURIComponent(redirectUrl)}`)
    return
  }

  // window.location.assign(`/`)
  // return
  try {
    console.debug('[adminAcrGuard] pass-through')
  } catch (_) {}
  next()
}

function getAvailableRoute(ability: Ability) {
  if (ability.can('read-all', 'Setting')) {
    return { name: 'admin-settings-general' }
  }

  if (ability.can('read-all', 'Account')) {
    return { name: 'admin-settings-users' }
  }

  if (ability.can('read-all', 'Group')) {
    return { name: 'admin-settings-groups' }
  }

  if (ability.can('read-all', 'Drive')) {
    return { name: 'admin-settings-spaces' }
  }

  throw Error('Insufficient permissions')
}

export const routes = ({ $ability }: { $ability: Ability }): RouteRecordRaw[] => [
  {
    path: '/',
    redirect: () => {
      return { name: 'admin-settings-general' }
    }
  },
  {
    path: '/general',
    name: 'admin-settings-general',
    component: General,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('General')
    }
  },
  {
    path: '/users',
    name: 'admin-settings-users',
    component: Users,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('Users')
    }
  },
  {
    path: '/groups',
    name: 'admin-settings-groups',
    component: Groups,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('Groups')
    }
  },
  {
    path: '/spaces',
    name: 'admin-settings-spaces',
    component: Spaces,
    beforeEnter: adminAcrGuard,
    meta: {
      authContext: 'user',
      title: $gettext('Spaces')
    }
  }
]

export const navItems = ({ $ability }: { $ability: Ability }): AppNavigationItem[] => [
  {
    name: $gettext('General'),
    icon: 'settings-4',
    route: {
      path: `/${appId}/general?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Setting')
    },
    priority: 10
  },
  {
    name: $gettext('Users'),
    icon: 'user',
    route: {
      path: `/${appId}/users?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Account')
    },
    priority: 20
  },
  {
    name: $gettext('Groups'),
    icon: 'group-2',
    route: {
      path: `/${appId}/groups?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Group')
    },
    priority: 30
  },
  {
    name: $gettext('Spaces'),
    icon: 'layout-grid',
    route: {
      path: `/${appId}/spaces?`
    },
    isVisible: () => {
      return $ability.can('read-all', 'Drive')
    },
    priority: 40
  }
]

export default defineWebApplication({
  setup() {
    const { can } = useAbility()
    const userStore = useUserStore()

    const appInfo: ApplicationInformation = {
      name: $gettext('Admin Settings'),
      id: appId,
      icon: 'settings-4',
      color: '#2b2b2b'
    }

    const menuItems = computed<AppMenuItemExtension[]>(() => {
      const items: AppMenuItemExtension[] = []

      const menuItemAvailable =
        userStore.user &&
        (can('read-all', 'Setting') ||
          can('read-all', 'Account') ||
          can('read-all', 'Group') ||
          can('read-all', 'Drive'))

      if (menuItemAvailable) {
        items.push({
          id: `app.${appInfo.id}.menuItem`,
          type: 'appMenuItem',
          label: () => appInfo.name,
          color: appInfo.color,
          icon: appInfo.icon,
          priority: 40,
          path: urlJoin(appInfo.id)
        })
      }

      return items
    })

    return {
      appInfo,
      routes,
      navItems,
      translations,
      extensions: menuItems
    }
  }
})
