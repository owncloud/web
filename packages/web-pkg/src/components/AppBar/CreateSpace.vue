<template>
  <oc-button
    id="new-space-menu-btn"
    key="new-space-menu-btn-enabled"
    v-oc-tooltip="showLabel ? undefined : $gettext('New space')"
    :aria-label="showLabel ? undefined : $gettext('New space')"
    appearance="filled"
    variation="primary"
    @click="showCreateSpaceModal"
  >
    <oc-icon name="add" />
    <span v-if="showLabel" v-text="$gettext('New Space')" />
  </oc-button>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { useGettext } from 'vue3-gettext'
import {
  useModals,
  useCreateSpace,
  useSpaceHelpers,
  useMessages,
  useSpacesStore,
  useResourcesStore
} from '../../composables'

export default defineComponent({
  props: {
    showLabel: {
      type: Boolean,
      default: true
    }
  },
  emits: ['spaceCreated'],
  setup(props, { emit }) {
    const { showMessage, showErrorMessage } = useMessages()
    const { $gettext } = useGettext()
    const { createSpace } = useCreateSpace()
    const { checkSpaceNameModalInput } = useSpaceHelpers()
    const { dispatchModal } = useModals()
    const spacesStore = useSpacesStore()
    const { upsertResource } = useResourcesStore()

    const addNewSpace = async (name: string) => {
      try {
        const createdSpace = await createSpace(name)
        upsertResource(createdSpace)
        spacesStore.upsertSpace(createdSpace)
        emit('spaceCreated', createdSpace)
        showMessage({ title: $gettext('Space was created successfully') })
      } catch (error) {
        console.error(error)
        showErrorMessage({
          title: $gettext('Creating space failedâ€¦'),
          errors: [error]
        })
      }
    }

    const showCreateSpaceModal = () => {
      dispatchModal({
        title: $gettext('Create a new space'),
        confirmText: $gettext('Create'),
        hasInput: true,
        inputLabel: $gettext('Space name'),
        inputValue: $gettext('New space'),
        onConfirm: (name: string) => addNewSpace(name),
        onInput: checkSpaceNameModalInput
      })
    }

    return { showCreateSpaceModal }
  }
})
</script>
