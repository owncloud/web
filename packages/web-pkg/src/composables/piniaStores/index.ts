export * from './apps'
export * from './auth'
export * from './capabilities'
export * from './clipboard'
export * from './config'
export * from './extensionRegistry'
export * from './messages'
export * from './modals'
export * from './resources'
export * from './shares'
export * from './spaces'
export * from './theme'
export * from './user'
export * from './webWorkers'
export * from './inviteTokensList'
export * from './app'
