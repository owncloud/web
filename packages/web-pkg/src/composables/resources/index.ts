export * from './useCanBeOpenedWithSecureView'
export * from './useGetResourceContext'
export * from './useLoadFileInfoById'
export * from './useResourceContents'
export * from './useResourcePath'
