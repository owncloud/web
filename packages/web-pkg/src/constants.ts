export const EVENT_TROW_MOUNTED = 'rowMounted'
export const EVENT_FILE_DROPPED = 'fileDropped'
export const EVENT_TROW_CONTEXTMENU = 'contextmenuClicked'

export abstract class ImageDimension {
  static readonly Thumbnail: [number, number] = [36, 36]
  static readonly Tile: [number, number] = [1000, 1000]
  static readonly Preview: [number, number] = [1200, 1200]
  static readonly Avatar: number = 64
}

export abstract class ImageType {
  static readonly Thumbnail: string = 'thumbnail'
  static readonly Preview: string = 'preview'
  static readonly Avatar: string = 'avatar'
}
