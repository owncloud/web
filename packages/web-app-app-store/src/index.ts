import translations from '../l10n/translations.json'
import { useGettext } from 'vue3-gettext'
import { computed, unref } from 'vue'
import {
  AppMenuItemExtension,
  defineWebApplication,
  Extension,
  useAbility,
  useUserStore
} from '@ownclouders/web-pkg'
import { urlJoin } from '@ownclouders/web-client'
import { RouteRecordRaw } from 'vue-router'
import { useRepositoriesStore } from './piniaStores'
import { AppStoreConfigSchema, AppStoreRepository } from './types'
import { APPID } from './appid'

export default defineWebApplication({
  setup({ applicationConfig }) {
    const { $gettext } = useGettext()
    const { can } = useAbility()
    const userStore = useUserStore()
    const repositoryStore = useRepositoriesStore()

    const defaultRepositories: AppStoreRepository[] = [
      {
        name: 'awesome-ocis',
        url: 'https://raw.githubusercontent.com/owncloud/awesome-ocis/main/webApps/apps.json'
      }
    ]
    if (applicationConfig?.repositories) {
      const { repositories } = AppStoreConfigSchema.parse(applicationConfig)
      repositoryStore.setRepositories(repositories || defaultRepositories)
    } else {
      repositoryStore.setRepositories(defaultRepositories)
    }

    const appInfo = {
      name: $gettext('App Store'),
      id: APPID,
      icon: 'store',
      color: '#ff6961'
    }

    const hasPermission = computed(() => {
      // TODO: which permission(s) do we need to check here?
      return userStore.user && can('read-all', 'Setting')
    })

    const routes: RouteRecordRaw[] = [
      {
        path: '/',
        name: 'root',
        component: () => import('./LayoutContainer.vue'),
        redirect: urlJoin(appInfo.id, 'list'),
        beforeEnter: (to, from, next) => {
          if (!unref(hasPermission)) {
            return next({ path: '/' })
          }
          next()
        },
        meta: {
          authContext: 'user'
        },
        children: [
          {
            path: 'list',
            name: 'list',
            component: () => import('./views/AppList.vue'),
            meta: {
              authContext: 'user',
              title: $gettext('App Store')
            }
          },
          {
            path: 'app/:appId',
            name: 'details',
            component: () => import('./views/AppDetails.vue'),
            meta: {
              authContext: 'user',
              title: $gettext('App Details')
            }
          }
        ]
      }
    ]

    const menuItemExtension: AppMenuItemExtension = {
      id: `app.${appInfo.id}.menuItem`,
      type: 'appMenuItem',
      label: () => appInfo.name,
      color: appInfo.color,
      icon: appInfo.icon,
      priority: 30,
      path: urlJoin(appInfo.id)
    }
    const extensions = computed(() => {
      const result: Extension[] = []

      if (unref(hasPermission)) {
        result.push(menuItemExtension)
      }

      return result
    })

    return {
      appInfo,
      routes,
      translations,
      extensions
    }
  }
})
