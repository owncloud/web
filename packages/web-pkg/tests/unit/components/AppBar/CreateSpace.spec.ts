import CreateSpace from '../../../../src/components/AppBar/CreateSpace.vue'
import { mockDeep } from 'vitest-mock-extended'
import { Resource, SpaceResource } from '@ownclouders/web-client'
import { defaultPlugins, mount, defaultComponentMocks } from '@ownclouders/web-test-helpers'
import { useMessages, useModals, useSpacesStore } from '../../../../src/composables'
import { unref } from 'vue'

const selectors = {
  newSpaceBtn: '#new-space-menu-btn'
}

describe('CreateSpace component', () => {
  it('should show the "New Space" button', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.find(selectors.newSpaceBtn).exists()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('should show a modal when clicking the "New Space" button', async () => {
    const { wrapper } = getWrapper()
    const { dispatchModal } = useModals()
    await wrapper.find(selectors.newSpaceBtn).trigger('click')
    expect(dispatchModal).toHaveBeenCalledTimes(1)
  })
  describe('method "addNewSpace"', () => {
    it('creates the space and updates the readme data after creation', async () => {
      const { wrapper, mocks } = getWrapper()
      const { modals } = useModals()
      await wrapper.find(selectors.newSpaceBtn).trigger('click')

      const graphMock = mocks.$clientService.graphAuthenticated
      const space = mockDeep<SpaceResource>()
      graphMock.drives.createDrive.mockResolvedValue(space)
      graphMock.drives.updateDrive.mockResolvedValue(space)
      mocks.$clientService.webdav.putFileContents.mockResolvedValue(mockDeep<Resource>())
      await unref(modals)[0].onConfirm('New Space')

      const spacesStore = useSpacesStore()
      expect(spacesStore.upsertSpace).toHaveBeenCalled()
      expect(wrapper.emitted('spaceCreated')).toBeTruthy()
    })
    it('shows a message when an error occurred', async () => {
      vi.spyOn(console, 'error').mockImplementation(() => undefined)
      const { wrapper, mocks } = getWrapper()
      const { modals } = useModals()
      await wrapper.find(selectors.newSpaceBtn).trigger('click')

      const graphMock = mocks.$clientService.graphAuthenticated
      graphMock.drives.createDrive.mockRejectedValue({})
      await unref(modals)[0].onConfirm('New Space')

      const { showErrorMessage } = useMessages()
      expect(showErrorMessage).toHaveBeenCalled()
    })
  })
})

function getWrapper() {
  const mocks = defaultComponentMocks()
  return {
    mocks,
    wrapper: mount(CreateSpace, {
      global: {
        mocks,
        provide: mocks,
        plugins: [...defaultPlugins({ piniaOptions: { stubActions: false } })]
      }
    })
  }
}
