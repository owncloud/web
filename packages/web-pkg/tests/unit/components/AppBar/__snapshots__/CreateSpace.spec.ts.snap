// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateSpace component > should show the "New Space" button 1`] = `
"<button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-primary oc-button-primary-filled" id="new-space-menu-btn">
  <!--v-if-->
  <!-- @slot Content of the button --> <span class="oc-icon oc-icon-m oc-icon-passive"><!----></span> <span>New Space</span>
</button>"
`;
